app/models/question.rb

```diff
@@ -7,4 +7,43 @@ class Question < ApplicationRecord
   belongs_to :created_by, class_name: 'User'
   belongs_to :updated_by, class_name: 'User'
   validates :question_type_id, presence: true
+  validates :version_independent_id, presence: true,
+                                     if: proc { |q| q.version > 1 }
+  validates :version, presence: true, uniqueness: { scope: :version_independent_id,
+                                                    message: 'versions should be unique across a revised response set' }
+
+  after_save :assign_version_independent_id,
+             if: proc { |q| q.version == 1 && q.version_independent_id.blank? }
+
+  def self.latest_versions
+    joins('INNER JOIN (SELECT version_independent_id, MAX(version) as version
+             FROM questions GROUP BY version_independent_id) q
+             ON q.version_independent_id = questions.version_independent_id
+             AND questions.version = q.version')
+  end
+
+  def assign_version_independent_id
+    if version == 1
+      update_attribute(:version_independent_id, "Q-#{id}")
+    else
+      # If you have a ResponseSet with a version greater than 1, it should
+      # already have a version_independent_id set.
+      raise ActiveRecord::RecordInvalid, self
+    end
+  end
+
+  def build_new_revision
+    new_revision = Question.new(content: content, created_at: created_at, updated_at: updated_at, created_by: created_by,
```

Comment:
Because I copied over all fields. 

The `updated_at` will get overwritten but in my mind revising something would mean it's not really a new object so I'd expect the `created_at` date to match the original.